### ISSUE TITLE
Tomcat Stacktrace is rendered as response

### ISSUE THREAD

**Issue by matusewicz** [zalando/problem-spring-web#84]
When we use problem-spring-web with tomcat servlet container and WITHOUT **https://github.com/zalando/logbook** we encounter the following behaviour:

Request with specified unsupported Accept header:

```
curl -v -H "Accept: text/xml" http://localhost:8080/hello
> GET /hello HTTP/1.1
> Host: localhost:8080
> Accept: text/xml
```
Results in the following response:
```
< HTTP/1.1 500
< Content-Type: text/html;charset=utf-8

<!DOCTYPE html><html><head><title>Apache Tomcat/8.5.5 - Error report</title><style type="text/css">H1 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:22px;} H2 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:16px;} H3 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:14px;} BODY {font-family:Tahoma,Arial,sans-serif;color:black;background-color:white;} B {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;} P {font-family:Tahoma,Arial,sans-serif;background:white;color:black;font-size:12px;}A {color : black;}A.name {color : black;}.line {height: 1px; background-color: #525D76; border: none;}</style> </head><body><h1>HTTP Status 500 - foo</h1><div class="line"></div><p><b>type</b> Exception report</p><p><b>message</b> <u>foo</u></p><p><b>description</b> <u>The server encountered an internal error that prevented it from fulfilling this request.</u></p><p><b>exception</b></p><pre>java.lang.IllegalArgumentException: foo
	de.foo.bar.api.Hello.hello(Hello.java:11)
	sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	java.lang.reflect.Method.invoke(Method.java:483)
	org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:221)
	org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:136)
	org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:114)
	org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:827)
	org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:738)
	org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85)
	org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:963)
	org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:897)
	org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
	org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:861)
	javax.servlet.http.HttpServlet.service(HttpServlet.java:622)
	org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
	javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	org.zalando.logbook.servlet.NormalStrategy.doFilter(NormalStrategy.java:42)
	org.zalando.logbook.servlet.LogbookFilter.doFilter(LogbookFilter.java:33)
	org.zalando.logbook.servlet.HttpFilter.doFilter(HttpFilter.java:32)
	org.springframework.boot.actuate.trace.WebRequestTraceFilter.doFilterInternal(WebRequestTraceFilter.java:105)
	org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
	org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
	org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:115)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:169)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	org.springframework.security.oauth2.provider.authentication.OAuth2AuthenticationProcessingFilter.doFilter(OAuth2AuthenticationProcessingFilter.java:176)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:121)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:66)
	org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56)
	org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
	org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177)
	org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:346)
	org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:262)
	org.zalando.logbook.servlet.SecurityStrategy.doFilter(SecurityStrategy.java:32)
	org.zalando.logbook.servlet.LogbookFilter.doFilter(LogbookFilter.java:33)
	org.zalando.logbook.servlet.HttpFilter.doFilter(HttpFilter.java:32)
	org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:197)
	org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	org.springframework.boot.actuate.autoconfigure.MetricsFilter.doFilterInternal(MetricsFilter.java:107)
	org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	org.zalando.tracer.servlet.TracerFilter.doFilter(TracerFilter.java:33)
	org.zalando.tracer.servlet.HttpFilter.doFilter(HttpFilter.java:28)
* Closing connection 0
</pre><p><b>note</b> <u>The full stack trace of the root cause is available in the Apache Tomcat/8.5.5 logs.</u></p><hr class="line"><h3>Apache Tomcat/8.5.5</h3></body></html>%
```

When you include **https://github.com/zalando/logbook** and enable trace logging 
``` <Logger name="org.zalando.logbook.Logbook" level="trace"/>```
the stack trace won't be rendered anymore inside the response and the response code changes from 500 to 406.

**Comment by matusewicz** [comment]
@lukasniemeier-zalando was investigating this issue as well and could provide more detailed information about the codepath that could cause this issue.

**Comment by lukasniemeier-zalando** [comment]
It looks like that our [current default implementation of the fallback](https://github.com/zalando/problem-spring-web/blob/master/src/main/java/org/zalando/problem/spring/web/advice/AdviceTrait.java#L161) does not close the response properly, as there is no body to write by Spring (`null`). Tomcat overwrites anything set by problem-spring-web by its own error page because the context contains an exception.

If you use logbook this `null` body is still touched/closed somehow, so Tomcat does run its error code. I will try to prepare a minimal example project.

**Comment by lukasniemeier-zalando** [comment]
What we should do as a default fallback is `response.sendError(406)` instead. This signals the servlet container that an error happened, but it's already handled.

**Comment by whiskeysierra** [comment]
Can we still send a response with that?

**Comment by whiskeysierra** [comment]
Doesn't sound like it:

> If the response has already been committed, this method throws an IllegalStateException. After using this method, the response should be considered to be committed and should not be written to.

**Comment by whiskeysierra** [comment]
@lukasniemeier-zalando Would an empty string as a body help here?

**Comment by lukasniemeier-zalando** [comment]
Yes I think this should work.

**Comment by whiskeysierra** [comment]
Would this technically be correct?

```java
final ResponseEntity<Problem> fallback = fallback(throwable, problem, request, headers);

if (fallback.getBody() == null) {
    final ServerHttpResponse response = new ServletServerHttpResponse(
            request.getNativeResponse(HttpServletResponse.class));

    response.setStatusCode(fallback.getStatusCode());
    response.getHeaders().putAll(fallback.getHeaders());

    try {
        response.getBody(); // just so we're actually flushing the body...
        response.flush();
    } catch (final IOException e) {
        throw new UncheckedIOException(e);
    }
}

return fallback;
```

**Comment by whiskeysierra** [comment]
I released 0.19.1
